File: cli-tool/python/route_planner.py
from typing import List, Dict, Optional, Tuple
from math import radians, sin, cos, sqrt, atan2
from dataclasses import dataclass


@dataclass
class Coordinate:
    """
    Represents a basic geographical coordinate.
    """
    lat: float
    lon: float


@dataclass
class NamedCoordinate(Coordinate):
    """
    Represents a coordinate with a name (used for stops in the route).
    """
    name: str


@dataclass
class MissionData:
    """
    Represents configuration settings for the route planning mission.
    """
    SPEED_M_PER_S: float
    SURVEY_TIME_S: int
    MAX_FLIGHT_TIME_S: int
    CHARGING_TIME_S: int
    MISSION_RADIUS_M: float


class RoutePlanner:
    """
    The RoutePlanner class calculates an optimal set of drone flight routes (legs)
    based on a list of coordinates and mission constraints such as speed, flight time, etc.
    """

    def __init__(self, mission_data: MissionData):
        """
        Initializes the planner with mission parameters.
        """
        self.SPEED_M_PER_S = mission_data.SPEED_M_PER_S
        self.SURVEY_TIME_S = mission_data.SURVEY_TIME_S
        self.MAX_FLIGHT_TIME_S = mission_data.MAX_FLIGHT_TIME_S
        self.CHARGING_TIME_S = mission_data.CHARGING_TIME_S
        self.MISSION_RADIUS_M = mission_data.MISSION_RADIUS_M

    def run_planner(self, coords: List[NamedCoordinate], custom_start: Optional[Coordinate] = None) -> Dict:
        """
        Executes the route planning algorithm and returns a structured result.

        :param coords: List of coordinates to survey
        :param custom_start: Optional starting point for the drone
        :return: Dictionary containing the planning result, route, and diagnostics
        """
        removed_points = []

        # Determine starting point
        if custom_start:
            start_point = custom_start
            updated_coords = [NamedCoordinate(lat=custom_start.lat, lon=custom_start.lon, name="Custom startpoint")] + coords
        else:
            avg_lat = sum(c.lat for c in coords) / len(coords)
            avg_lon = sum(c.lon for c in coords) / len(coords)
            start_point = Coordinate(avg_lat, avg_lon)
            updated_coords = [NamedCoordinate(lat=avg_lat, lon=avg_lon, name="Calculated best dock location")] + coords

        # Filter out coordinates beyond allowed mission radius
        updated_coords = [
            c for i, c in enumerate(updated_coords)
            if i == 0 or self.haversine_distance(start_point, c) <= self.MISSION_RADIUS_M or removed_points.append(c)
        ]

        start_idx = 0
        distance_matrix = self.create_distance_matrix(updated_coords)
        max_attempts = len(updated_coords) - 1

        # Try route planning, remove unreachable points iteratively if needed
        while max_attempts > 0:
            try:
                route = self.plan_full_route(start_idx, updated_coords, distance_matrix)
                duration, distance = self.get_route_duration_and_distance(route, updated_coords, start_idx, distance_matrix)
                output = self.build_route_output(route, updated_coords, start_idx, distance_matrix)

                return {
                    "success": True,
                    "startPoint": updated_coords[start_idx],
                    "totalDuration": duration,
                    "totalDistance": distance,
                    "removedPoints": removed_points,
                    "legs": output
                }
            except Exception:
                candidates = updated_coords[1:]
                found = False

                for candidate in candidates:
                    test_coords = [c for c in updated_coords if c != candidate]
                    test_matrix = self.create_distance_matrix(test_coords)
                    try:
                        self.plan_full_route(start_idx, test_coords, test_matrix)
                        updated_coords = test_coords
                        distance_matrix = test_matrix
                        removed_points.append(candidate)
                        found = True
                        break
                    except Exception:
                        continue

                if not found:
                    return {
                        "success": False,
                        "error": "No valid route possible after removing all candidates.",
                        "removedPoints": removed_points
                    }

            max_attempts -= 1

        return {
            "success": False,
            "error": "There is no possible route.",
            "removedPoints": removed_points
        }

    def haversine_distance(self, a: Coordinate, b: Coordinate) -> float:
        """
        Calculates the haversine distance (in meters) between two geographic points.
        """
        R = 6371000  # Earth radius in meters
        d_lat = radians(b.lat - a.lat)
        d_lon = radians(b.lon - a.lon)
        lat1 = radians(a.lat)
        lat2 = radians(b.lat)

        a_calc = sin(d_lat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(d_lon / 2) ** 2
        c = 2 * atan2(sqrt(a_calc), sqrt(1 - a_calc))
        return R * c

    def create_distance_matrix(self, coords: List[Coordinate]) -> List[List[float]]:
        """
        Creates a 2D matrix of distances between all coordinate pairs.

        :param coords: List of coordinates
        :return: Distance matrix
        """
        return [
            [0 if i == j else self.haversine_distance(coords[i], coords[j]) for j in range(len(coords))]
            for i in range(len(coords))
        ]

    def compute_travel_time(self, distance: float) -> float:
        """
        Converts a distance (in meters) to travel time (in seconds).
        """
        return distance / self.SPEED_M_PER_S

    def can_add_to_leg(self, current_time: float, travel_time: float, return_time: float) -> bool:
        """
        Checks whether a point can be added to the current leg
        without exceeding the drone's max flight time.
        """
        return current_time + travel_time + self.SURVEY_TIME_S + return_time <= self.MAX_FLIGHT_TIME_S

    def plan_leg(self, start_idx: int, unvisited: set, distance_matrix: List[List[float]]) -> List[int]:
        """
        Plans a single leg of the flight using a greedy approach.

        :param start_idx: Index of the dock/start coordinate
        :param unvisited: Set of remaining point indices
        :param distance_matrix: Precomputed distance matrix
        :return: List of point indices visited in this leg
        """
        leg = []
        time_used = 0
        current = start_idx

        while True:
            next_point = None
            shortest_time = float("inf")

            for idx in unvisited:
                travel_time = self.compute_travel_time(distance_matrix[current][idx])
                return_time = self.compute_travel_time(distance_matrix[idx][start_idx])

                if self.can_add_to_leg(time_used, travel_time, return_time) and travel_time < shortest_time:
                    next_point = idx
                    shortest_time = travel_time

            if next_point is None:
                break

            travel = self.compute_travel_time(distance_matrix[current][next_point])
            time_used += travel + self.SURVEY_TIME_S
            leg.append(next_point)
            unvisited.remove(next_point)
            current = next_point

        return leg

    def plan_full_route(self, start_idx: int, coords: List[NamedCoordinate], distance_matrix: List[List[float]]) -> List[List[int]]:
        """
        Plans the full route including multiple legs until all points are visited.

        :raises: Exception if a valid route cannot be completed
        """
        unvisited = set(range(len(coords)))
        unvisited.remove(start_idx)
        route = []

        while unvisited:
            leg = self.plan_leg(start_idx, unvisited, distance_matrix)
            if not leg:
                raise Exception("Some points are not reachable.")
            route.append(leg)

        return route

    def get_route_duration_and_distance(
        self,
        route: List[List[int]],
        coords: List[NamedCoordinate],
        start_idx: int,
        distance_matrix: List[List[float]]
    ) -> Tuple[float, float]:
        """
        Calculates total flight time and distance across all legs.

        :return: Tuple (total_duration_in_seconds, total_distance_in_meters)
        """
        total_time = 0
        total_distance = 0
        current = start_idx

        for leg in route:
            for next_idx in leg:
                dist = distance_matrix[current][next_idx]
                total_time += self.compute_travel_time(dist) + self.SURVEY_TIME_S
                total_distance += dist
                current = next_idx

            return_dist = distance_matrix[current][start_idx]
            total_time += self.compute_travel_time(return_dist) + self.CHARGING_TIME_S
            total_distance += return_dist
            current = start_idx

        return total_time, total_distance

    def build_route_output(
        self,
        route: List[List[int]],
        coords: List[NamedCoordinate],
        start_idx: int,
        distance_matrix: List[List[float]]
    ) -> List[Dict]:
        """
        Builds the final structured output for all legs and stops.

        :return: List of dictionaries representing each leg
        """
        output = []
        current = start_idx
        index = 1
        global_stop_index = 1

        for leg in route:
            leg_distance = 0
            leg_time = 0
            stops = []

            for next_idx in leg:
                dist = distance_matrix[current][next_idx]
                time = self.compute_travel_time(dist)
                leg_distance += dist
                leg_time += time + self.SURVEY_TIME_S

                stops.append({
                    "stopIndex": global_stop_index,
                    "name": coords[next_idx].name,
                    "lat": coords[next_idx].lat,
                    "lon": coords[next_idx].lon,
                    "distance": dist,
                    "flightTime": time,
                    "surveyTime": self.SURVEY_TIME_S
                })

                global_stop_index += 1
                current = next_idx

            return_dist = distance_matrix[current][start_idx]
            return_time = self.compute_travel_time(return_dist)
            leg_distance += return_dist
            leg_time += return_time + self.CHARGING_TIME_S

            stops.append({
                "stopIndex": global_stop_index,
                "name": "Dock",
                "lat": coords[start_idx].lat,
                "lon": coords[start_idx].lon,
                "distance": return_dist,
                "flightTime": return_time,
                "chargingTime": self.CHARGING_TIME_S
            })

            global_stop_index += 1
            output.append({
                "legIndex": index,
                "duration": leg_time,
                "distance": leg_distance,
                "returnToHome": {
                    "distance": return_dist,
                    "flightTime": return_time,
                    "chargingTime": self.CHARGING_TIME_S
                },
                "stops": stops
            })

            current = start_idx
            index += 1

        return output

File: server/src/routePlanner.ts
import type { MissionData } from "./interfaces/missiondata.interface";
import type { Coordinate, NamedCoordinate } from "./interfaces/coordinate.interface";

/**
 * The routePlanner class calculates optimized drone routes and the perfect starting point.
 * It accounts for multiple variables.
 * A route may include multiple legs (flights), each constrained by the drone's endurance.
 */
export class routePlanner {
    // Default constants 
    SPEED_M_PER_S: number = 15; // Flight speed in meters per second
    SURVEY_TIME_S: number = 600; // Time spent surveying each point (in seconds)
    MAX_FLIGHT_TIME_S: number = 2400; // Maximum flight duration before return (in seconds)
    CHARGING_TIME_S: number = 2100; // Charging or break time between legs (in seconds)
    MISSION_RADIUS_M: number = 10000; // Max distance from start point (in meters)
    CUSTOM_START_CORD: Coordinate = null; // Optional fixed start coordinate

    constructor(missionData: MissionData) {
        this.SPEED_M_PER_S = missionData.SPEED_M_PER_S;
        this.SURVEY_TIME_S = missionData.SURVEY_TIME_S;
        this.MAX_FLIGHT_TIME_S = missionData.MAX_FLIGHT_TIME_S;
        this.CHARGING_TIME_S = missionData.CHARGING_TIME_S;
        this.MISSION_RADIUS_M = missionData.MISSION_RADIUS_M;
    }

    /**
     * Main entry point: Computes the full drone route plan.
     * 
     * @param coords - Array of target coordinates with names
     * @param customStart - Optional custom start coordinate
     * @returns Route result containing legs, removed points, total duration, etc.
     */
    public runPlanner(coords: NamedCoordinate[], customStart?: Coordinate): any {
        let updatedCoords: NamedCoordinate[];
        let startIdx = 0;
        let route: number[][];
        let duration: number;
        let distance: number;
        let startPoint: Coordinate;
        let removedPoints: NamedCoordinate[] = [];

        // Determine start point: use custom if provided, otherwise calculate centroid
        if (customStart) {
            startPoint = customStart;
            updatedCoords = [{ name: "Custom startpoint", ...customStart }, ...coords];
        } else {
            const centroid = coords.reduce(
                (acc, c) => ({ lat: acc.lat + c.lat, lon: acc.lon + c.lon }),
                { lat: 0, lon: 0 }
            );
            centroid.lat /= coords.length;
            centroid.lon /= coords.length;
            startPoint = centroid;
            updatedCoords = [{ name: "Calculated best dock location", ...centroid }, ...coords];
        }

        // Filter out points that exceed the mission radius from the start point
        updatedCoords = updatedCoords.filter((coord, idx) => {
            const dist = this.haversineDistance(startPoint, coord);
            if (dist > this.MISSION_RADIUS_M && idx !== 0) {
                removedPoints.push(coord);
                return false;
            }
            return true;
        });

        let distanceMatrix = this.createDistanceMatrix(updatedCoords);
        let maxAttempts = updatedCoords.length - 1;

        // Try to build a valid route; fallback by removing unreachable points one by one
        while (maxAttempts-- > 0) {
            try {
                route = this.planFullRoute(startIdx, updatedCoords, distanceMatrix);
                const result = this.getRouteDurationAndDistance(route, updatedCoords, startIdx, distanceMatrix);
                duration = result.duration;
                distance = result.distance;
                break;
            } catch (e) {
                const candidates = updatedCoords.slice(1);
                let found = false;

                for (const candidate of candidates) {
                    const testCoords = updatedCoords.filter(c => c !== candidate);
                    const testMatrix = this.createDistanceMatrix(testCoords);

                    try {
                        this.planFullRoute(startIdx, testCoords, testMatrix);
                        updatedCoords = testCoords;
                        removedPoints.push(candidate);
                        distanceMatrix = testMatrix;
                        found = true;
                        break;
                    } catch (_) {
                        continue;
                    }
                }

                if (!found) {
                    return {
                        success: false,
                        error: "No valid route possible after removing all candidates.",
                        removedPoints
                    };
                }
            }
        }

        if (!route) {
            return {
                success: false,
                error: "There is no possible route.",
                removedPoints
            };
        }

        const output = this.buildRouteOutput(route, updatedCoords, startIdx, distanceMatrix);

        return {
            success: true,
            startPoint: updatedCoords[startIdx],
            totalDuration: duration,
            totalDistance: distance,
            removedPoints,
            legs: output
        };
    }

    /** Converts degrees to radians */
    private toRad(deg: number): number {
        return (deg * Math.PI) / 180;
    }

    /** Calculates haversine distance (in meters) between two GPS coordinates */
    private haversineDistance(a: Coordinate, b: Coordinate): number {
        const R = 6371e3; // Earth's radius in meters
        const dLat = this.toRad(b.lat - a.lat);
        const dLon = this.toRad(b.lon - a.lon);
        const lat1 = this.toRad(a.lat);
        const lat2 = this.toRad(b.lat);

        const aCalc =
            Math.sin(dLat / 2) ** 2 +
            Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;

        const c = 2 * Math.atan2(Math.sqrt(aCalc), Math.sqrt(1 - aCalc));
        return R * c;
    }

    /** Generates a matrix with pairwise distances between all coordinates */
    private createDistanceMatrix(coords: Coordinate[]): number[][] {
        const matrix: number[][] = Array.from({ length: coords.length }, () => new Array(coords.length).fill(0));
        for (let i = 0; i < coords.length; i++) {
            for (let j = 0; j < coords.length; j++) {
                matrix[i][j] = i === j ? 0 : this.haversineDistance(coords[i], coords[j]);
            }
        }
        return matrix;
    }

    /** Converts distance into flight time */
    private computeTravelTime(distance: number): number {
        return distance / this.SPEED_M_PER_S;
    }

    /** Checks if a point can be added to a leg without exceeding max flight time */
    private canAddToLeg(currentTime: number, travelTime: number, returnTime: number): boolean {
        return currentTime + travelTime + this.SURVEY_TIME_S + returnTime <= this.MAX_FLIGHT_TIME_S;
    }

    /**
     * Plans one flight leg by greedily visiting as many reachable points as possible
     */
    private planLeg(startIdx: number, unvisited: Set<number>, distanceMatrix: number[][]): number[] {
        const leg: number[] = [];
        let timeUsed = 0;
        let current = startIdx;

        while (true) {
            let next: number | null = null;
            let shortestTime = Infinity;

            for (const idx of unvisited) {
                const travelTime = this.computeTravelTime(distanceMatrix[current][idx]);
                const returnTime = this.computeTravelTime(distanceMatrix[idx][startIdx]);

                if (this.canAddToLeg(timeUsed, travelTime, returnTime) && travelTime < shortestTime) {
                    next = idx;
                    shortestTime = travelTime;
                }
            }

            if (next === null) break;

            const travel = this.computeTravelTime(distanceMatrix[current][next]);
            timeUsed += travel + this.SURVEY_TIME_S;
            leg.push(next);
            unvisited.delete(next);
            current = next;
        }

        return leg;
    }

    /**
     * Plans a complete route consisting of multiple flight legs
     * Throws an error if some points are unreachable
     */
    private planFullRoute(startIdx: number, coords: NamedCoordinate[], distanceMatrix: number[][]): number[][] {
        const unvisited = new Set<number>(coords.map((_, i) => i));
        unvisited.delete(startIdx);

        const route: number[][] = [];

        while (unvisited.size > 0) {
            const leg = this.planLeg(startIdx, unvisited, distanceMatrix);
            if (leg.length === 0) throw new Error("Some points are not reachable.");
            route.push(leg);
        }

        return route;
    }

    /** Calculates total route time and distance, including returns and charging times */
    private getRouteDurationAndDistance(
        route: number[][],
        coords: NamedCoordinate[],
        startIdx: number,
        distanceMatrix: number[][]
    ): { duration: number, distance: number } {
        let totalTime = 0;
        let totalDistance = 0;
        let current = startIdx;

        for (const leg of route) {
            for (const next of leg) {
                const dist = distanceMatrix[current][next];
                const toNext = this.computeTravelTime(dist);
                totalTime += toNext + this.SURVEY_TIME_S;
                totalDistance += dist;
                current = next;
            }

            const toBase = distanceMatrix[current][startIdx];
            const toBaseTime = this.computeTravelTime(toBase);
            totalTime += toBaseTime + this.CHARGING_TIME_S;
            totalDistance += toBase;
            current = startIdx;
        }

        return { duration: totalTime, distance: totalDistance };
    }

    /**
     * Converts a raw route into a structured output format containing legs and stops
     */
    private buildRouteOutput(
        route: number[][],
        coords: NamedCoordinate[],
        startIdx: number,
        distanceMatrix: number[][]
    ) {
        const output = [];
        let current = startIdx;
        let index = 1;
        let globalStopIndex = 1;

        for (const leg of route) {
            let legDistance = 0;
            let legTime = 0;
            const stops = [];

            for (const next of leg) {
                const dist = distanceMatrix[current][next];
                const time = this.computeTravelTime(dist);
                legDistance += dist;
                legTime += time + this.SURVEY_TIME_S;

                stops.push({
                    stopIndex: globalStopIndex++,
                    name: coords[next].name,
                    lat: coords[next].lat,
                    lon: coords[next].lon,
                    distance: dist,
                    flightTime: time,
                    surveyTime: this.SURVEY_TIME_S
                });

                current = next;
            }

            const returnDist = distanceMatrix[current][startIdx];
            const returnTime = this.computeTravelTime(returnDist);
            legDistance += returnDist;
            legTime += returnTime + this.CHARGING_TIME_S;

            stops.push({
                stopIndex: globalStopIndex++,
                name: "Dock",
                lat: coords[startIdx].lat,
                lon: coords[startIdx].lon,
                distance: returnDist,
                flightTime: returnTime,
                chargingTime: this.CHARGING_TIME_S
            });

            output.push({
                legIndex: index++,
                duration: legTime,
                distance: legDistance,
                returnToHome: {
                    distance: returnDist,
                    flightTime: returnTime,
                    chargingTime: this.CHARGING_TIME_S
                },
                stops
            });

            current = startIdx;
        }

        return output;
    }
}
